File: Simulacao/MM_2/atores/operador.js
class Operadores {

    constructor(quantidade) {
        this.operadores = [];
        var i;

        for (i = 0; i < quantidade; i++) {
            this.operadores.push(new Operador());
        }
    }

    verOperadorLivre() {
        const index = this.operadores.findIndex(operador => operador.ES == 0);

        if (index >= 0) {
            return true;
        }
        return false;
    }

    ocuparOperador(TR) {
        const index = this.operadores.findIndex(operador => operador.ES == 0);

        if (index >= 0) {
            this.operadores[index].changeES();
            this.operadores[index].addTempoLivre(TR);
        }
        return index;
    }

    aliviarOperador(TR, index) {
        if (this.operadores[index].ES == 1) {
            this.operadores[index].changeES();
            this.operadores[index].setLastTR(TR);
        }
    }

    sistemaNaoVazio() {
        const index = this.operadores.findIndex(operador => operador.ES == 1);

        if (index >= 0) {
            return true;
        }
        return false;
    }

}

class Operador {

    constructor() {
        this.ES = 0;
        this.lastTR = 0;
        this.tempoLivre = 0;
    }

    changeES() {
        this.ES = this.ES == 1 ? 0 : 1;
    }

    setLastTR(TR) {
        this.lastTR = TR
    }

    addTempoLivre(TR) {
        this.tempoLivre += (TR - this.lastTR);
    }
}

File: Simulacao/MM2_Python/Operador.py
class Operador:
    def __init__(self, quantidade):
        self.operadores = []
        for i in range(0, quantidade):
            self.operadores.append(operadores())

    def verOperadorLivre(self):
        for x in self.operadores:
            if x.ES == 0:
                return 1
        return 0
    def ocuparOperador(self, TR):
        for x in range(len(self.operadores)):
            if self.operadores[x].ES == 0:
                self.operadores[x].changeES()
                self.operadores[x].addTempoLivre(TR)
                return x
    def aliviarOperador(self, TR, index):
        if self.operadores[index].ES == 1:
            self.operadores[index].changeES()
            self.operadores[index].setLastTR(TR)
    def sistemNotVazio(self):
        for x in self.operadores:
            if x.ES == 1:
                return 1

    #def printEstado(self):
    #    for x in self.operadores:
    #        print('ES: ' + repr(x.ES) + ' lastTR: ' + repr(x.lastTR) + ' tempoLivre: ' + repr(x.tempoLivre))
        
class operadores:
    def __init__(self):
        self.ES = 0
        self.lastTR = 0
        self.tempoLivre = 0
    def changeES(self):
        if self.ES == 1:
            self.ES = 0
        else:
            self.ES = 1
    def setLastTR(self, TR):
        self.lastTR = TR

    def addTempoLivre(self, TR):
        self.tempoLivre = self.tempoLivre + (TR - self.lastTR)